src/main/java/org/elasticsearch/index/query/SimpleQueryStringBuilder.java

```diff
@@ -373,32 +364,171 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
+            builder.field("lenient", lenient);
+        }
+
+        if (analyzeWildcard != null) {
+            builder.field("analyze_wildcard", analyzeWildcard);
+        }
+
+        if (locale != null) {
+            builder.field("locale", locale.toString());
         }
 
-        Query query = sqp.parse(queryBody);
         if (queryName != null) {
-            parseContext.addNamedQuery(queryName, query);
+            builder.field("_name", queryName);
         }
 
-        if (minimumShouldMatch != null && query instanceof BooleanQuery) {
-            Queries.applyMinimumShouldMatch((BooleanQuery) query, minimumShouldMatch);
+        if (minimumShouldMatch != null) {
+            builder.field("minimum_should_match", minimumShouldMatch);
         }
-        return query;
+
+        builder.endObject();
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        queryText = in.readOptionalString();
+        int size = in.readInt();
+        for (int i = 0; i < size; i++) {
+            String field = in.readString();
+            Float weight = in.readFloat();
+            fieldsAndWeights.put(field, weight);
+        }
+        flags = in.readInt();
+        analyzerName = in.readOptionalString();
+        String opStr = in.readOptionalString();
+        if ("and".equals(opStr)) {
+            defaultOperator = Operator.AND;
+        }
+        if ("or".equals(opStr)) {
+            defaultOperator = Operator.OR;
+        }
+        lowercaseExpandedTerms = in.readOptionalBoolean();
+        lenient = in.readOptionalBoolean();
+        analyzeWildcard = in.readOptionalBoolean();
+        if (in.readBoolean()) {
+            String localeStr = in.readString();
+            locale = LocaleUtils.parse(localeStr);
+        }
+        queryName = in.readOptionalString();
+        minimumShouldMatch = in.readOptionalString();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(queryText);
+        out.writeInt(fieldsAndWeights.size());
+        for (Map.Entry<String, Float> entry : fieldsAndWeights.entrySet()) {
+            out.writeString(entry.getKey());
+            out.writeFloat(entry.getValue());
+        }
+        out.writeInt(flags);
+        out.writeOptionalString(analyzerName);
+        out.writeBoolean(defaultOperator != null);
+        if (defaultOperator != null) {
+            out.writeString(defaultOperator.name().toLowerCase(Locale.ROOT));
+        }
+        out.writeOptionalBoolean(lowercaseExpandedTerms);
+        out.writeOptionalBoolean(lenient);
+        out.writeOptionalBoolean(analyzeWildcard);
+        out.writeBoolean(locale != null);
+        if (locale != null) {
+            out.writeString(locale.toString());
+        }
+        out.writeOptionalString(queryName);
+        out.writeOptionalString(minimumShouldMatch);
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = super.hashCode();
```

Comment:
I have a question here since we will have to do this in other classes too. Calling super.hashCode() will call Object.hashCode() if we don't implement hashCode() also in BaseQueryBuilder. If so, can it happen that I have two query instances that are equal by means of equals() but have a different hashCode()? I always get confused with this, just asking out of curiosity.